### ISSUE TITLE
[HOLD #11768] [$16000] Desktop - The Back ⌘[ and Forward ⌘] shortcuts keys are not working as expected in LHN

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#4612]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
This issue has been split in two. This one is solely focused on the back/forward actions in LeftHandNav (LHN), the other https://github.com/Expensify/App/issues/8657 is focused on back/forward actions in the main chat window.

## Action Performed:
1. Open desktop app
2. Navigate to several conversations
3. Use the back `⌘[` and Forward `⌘]` shortcuts

## Expected Result:
Back `⌘[` and Forward `⌘]` shortcuts should work and navigate through the previously navigated conversations and show correctly in LHN.

## Actual Result:
Shortcuts are not working as expected. 

## Workaround:
User has to manually navigate through the conversations. 

## Platform:

Where is this issue occurring?

- Desktop App

**Version Number:** 1.0.82-7

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation
Unable to get a video atm. Will update with video later. 
Other issue for back/forward working correctly in the main chat window https://github.com/Expensify/App/issues/8657 

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___ 
From @mallenexpensify https://expensify.slack.com/archives/C01GTK53T8Q/p1628722621224600
> The Back ⌘[ and Forward ⌘] shortcuts keys don't work properly on Desktop Version 1.0.82-7

**Comment by MelvinBot** [comment]
Triggered auto assignment to @joelbettner (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
@joelbettner Eep! 4 days overdue now. Issues have feelings too...

**Comment by mallenexpensify** [comment]
@joelbettner , do you think this can be worked on by a contributor? If so, can you add the `External` label? Thx

**Comment by joelbettner** [comment]
Sorry, I've been nose down in allocations. Yes, I think this can be worked on by a contributor.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MitchExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by parasharrajat** [comment]
### Proposal
Here we have two options to configure this. 
1. the First Option is to register the keyboard short to with these keys to navigate back/forth in history using the `History API` 's back and forward methods.

for this, we have `KeyboardShortcut` lib. https://github.com/Expensify/App/blob/2b57f0fac59d007441fb170fd05445d83392ed86/src/libs/KeyboardShortcut/index.js

but we only have to activate these shortcuts for the desktop.

2. the second option is to handle these shortcuts from the Main.js desktop file, it is a little involved but does the same this. 
First, we have to tap the keyboard events from the browserWindow using `before-input-event` event.
```
win.webContents.on('before-input-event', (event, input) 
```

Then we have to call  `web-contents#contentsgoback` and `web-contents#contentsgoforward` to navigate. https://www.electronjs.org/docs/api/web-contents#contentsgoback

I think the first one is going to be easier.

**Comment by MitchExpensify** [comment]
[Upwork job](https://www.upwork.com/jobs/~01a10d83f7e4a593b9)

**Comment by parasharrajat** [comment]
@MitchExpensify I think this needs Exported label so that an engineer can review proposals.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NikkiWines (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by NikkiWines** [comment]
@parasharrajat's proposal (Option 1) looks good, @MitchExpensify please hire them for this issue! 🙇

**Comment by MitchExpensify** [comment]
Hired!

**Comment by parasharrajat** [comment]
I am still have to dig more into this. I will update soon

**Comment by kadiealexander** [comment]
Hi @parasharrajat, I've placed this issue on hold as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1631754482374800), as we are prioritising issues related to a feature release scheduled for 9/24. As an apology for the delay, we will add a $100 bonus as a thank you for waiting.

**Comment by NikkiWines** [comment]
MevlinBot this isn't overdue

**Comment by NikkiWines** [comment]
Still not overdue

**Comment by NikkiWines** [comment]
(changing the label for now so that MelvinBot stops adding the overdue label as frequently)

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by NikkiWines** [comment]
@parasharrajat any update on this or should we open this back up for other contributors to take on?

**Comment by parasharrajat** [comment]
I am sorry, I will give it a go this weekend and update you by Monday,

**Comment by parasharrajat** [comment]
Ok. So I am looking into this. I think the changes I did https://github.com/Expensify/App/pull/6207 here are also needed for this to work. 
The main issue is that history is not created correctly when we use chat switcher to change the chats. I will test it on that branch and share the updates shortly.

**Comment by NikkiWines** [comment]
Looks like the review for https://github.com/Expensify/App/pull/6207 is on hold while Marc is OOO, so seemsl like we should expect any progress on this to be delayed until next week

**Comment by marcaaron** [comment]
Does this feel like a potential `react-navigation` problem? 

>history is not created correctly when we use chat switcher to change the chats

I'd be curious to see if there are any new developments with the web implementation of `react-navigation`. IIRC a lot of the custom logic I added this past year to fix the history was basically just hacks to get browser history to work correctly and wasn't intended to be a long term solution.

The way I see it, we _could_ double down on the custom navigation action path we went down - but also try to see if we can influence a change in `react-navigation`. Maybe there's an open issue we can add context to or create a new one ([web issues]( https://github.com/react-navigation/react-navigation/issues?q=is%3Aopen+is%3Aissue+label%3Aplatform%3Aweb )).

**Comment by parasharrajat** [comment]
I have recently updated the `react-navigation` to the latest so maybe retest this issue and try out default behaviour.

**Comment by marcaaron** [comment]
>so maybe retest this issue and try out default behaviour

Ok so, who should do this? 😄

**Comment by parasharrajat** [comment]
**I**  will do this.

**Comment by parasharrajat** [comment]
Updated are posted on #6207 a couple of days back. Waiting for the response there.

**Comment by NikkiWines** [comment]
@parasharrajat any update on this one?

**Comment by parasharrajat** [comment]
Oops, forgot to check it due to hold on 6207 but I will look into it from now on. It could take a day or two.

**Comment by parasharrajat** [comment]
@mallenexpensify Could you please retest it may have been resolved? I can't reproduce it. The fact that I changed the navigation in https://github.com/Expensify/App/pull/6207 to work the same as LHN and this issue not reproducible via LHN says that it is fixed.

**Comment by mallenexpensify** [comment]
@parasharrajat 
> The Back ⌘[ and Forward ⌘] shortcuts keys are not working as expected
I'm still able reproduce the issue on Desktop Version 1.1.18-3.   

1. Via search/chat switcher - Jack > Monte > Deeter > Manager.  
2. `CMD + [` goes to Joe G.  
3. `CMD + ]` goes to Manager. 

Via clicking on user names in LHN seems to work _better_, I had one time where it navigated correctly using `CMD + [` but also had other times/tests where it didn't work correctly

**Comment by mallenexpensify** [comment]
@parasharrajat can you do more testing based on my comment above and see what you find? I'm still on Version 1.1.18-3

**Comment by parasharrajat** [comment]
Ok, I think I found a strange issue similar to what you are facing. this is reproducible from LHN as well. It seems more severe from what you are facing. But unfortunately, I haven't got a solution for that and nor time to analyze this ATM.

I request to open this to the pool.

**Comment by mallenexpensify** [comment]
Want me to leave you assigned as Contributor-Plus @parasharrajat ?

**Comment by parasharrajat** [comment]
Yup, that's fine. I have some context to this problem.

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the Exported assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @NikkiWines is eligible for the Exported assigner, not assigning anyone new.

**Comment by mallenexpensify** [comment]
Price doubled! I need to update the Upwork job, they're having some issues.

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the Exported assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @AndrewGable (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mallenexpensify** [comment]
Doubled price to $2000 https://www.upwork.com/jobs/~01a4e8881ff6937bae
This job is open to proposals, @parasharrajat is currently assigned as the contributor-plus

**Comment by mallenexpensify** [comment]
Doubled price to $4000 https://www.upwork.com/jobs/~01a4e8881ff6937bae

**Comment by murataka** [comment]
Hello,

Currently the shourtcuts seem to be not configured in this version already :

### CONST.js

  ```
KEYBOARD_SHORTCUTS: {
        SEARCH: {
            descriptionKey: 'search',
            shortcutKey: 'K',
            modifiers: ['CTRL'],
        },
        NEW_GROUP: {
            descriptionKey: 'newGroup',
            shortcutKey: 'K',
            modifiers: ['CTRL', 'SHIFT'],
        },
        SHORTCUT_MODAL: {
            descriptionKey: 'openShortcutDialog',
            shortcutKey: 'I',
            modifiers: ['CTRL'],
        },
        ESCAPE: {
            descriptionKey: 'escape',
            shortcutKey: 'Escape',
            modifiers: [],
        },
        ENTER: {
            descriptionKey: null,
            shortcutKey: 'Enter',
            modifiers: [],
        },
        COPY: {
            descriptionKey: 'copy',
            shortcutKey: 'C',
            modifiers: ['CTRL'],
        },
    },
```

Is someone working on it and does some older version include those configuration ? @parasharrajat  I see that you already worked on that. I think I can solve that issue.

**Comment by parasharrajat** [comment]
We are open to proposals. Great to hear you can solve it. Let's see your approach.

**Comment by murataka** [comment]
> 

Thanks , I expect that there's an existing configuration somewhere , do you remember if there's more complete configuration of that file ? And sometimes the shortcut keys will overlap with browser shortcuts, shall we stick to previously defined shortcuts and cancel the browser shortcuts where essential ?

**Comment by parasharrajat** [comment]
These shortcuts are related to the Desktop app so they are configured on desktop/main.js

**Comment by murataka** [comment]
> These shortcuts are related to the Desktop app so they are configured on desktop/main.js

Thanks for that information,

I think we can move all configuration to one file , because it seems better to be able to use same shortcuts on native or browser. 

Let me know if this is a good Idea, and I shall have a look if there will be an issue implementing it that way ( if it is possible to make it work same on native and web ) if you think so .

**Comment by parasharrajat** [comment]
So CONST.js file lives on the renderer process and Desktop shortcuts on the main process. Apart from it, desktop shortcuts do not use listeners they utilize native Key bindings. It won't be possible atm to have the same config file. 

The shortcuts are configured correctly both for web and desktop. The main problem lies in the feature that they are bound to. The feature is not implemented/configured correctly.

**Comment by murataka** [comment]
> So CONST.js file lives on the renderer process and Desktop shortcuts on the main process.

I would like to hear the  expected behavior regardless of the technical difficulties ( to make it more clear ). Can of course leave it as is, so again , do you think we should try to enable those shortcut keys on browser also ? 


> The shortcuts are configured correctly both for web and desktop. The main problem lies in the feature that they are bound to. That feature is not implemented/configured correctly.

Can still leave them in separate configuration files , so we can later decide to make changes according to the intended behavior.

**Comment by murataka** [comment]
### Proposal

I realised that  `⌘[` and Forward `⌘]`  keys correcly call the goBack and goForward functions, however , that functionality seems to be broken : 



1. On firefox , back and forward does nothing
2. On Chrome , forward and back buttons work, but after doing back -> forward -> back -> forward , we get ```
useLinking.js:33 Uncaught TypeError: Cannot read property 'id' of undefined
    at useLinking.js:33
``` and no more history navigation possible.
3. On native , the experience is same with chrome.

> Back `⌘[` and Forward `⌘]` shortcuts should work and navigate through the previously navigated conversations

if this is the case ,

When we navigate to https://staging.new.expensify.com/settings for example , that is also pushed to the navigation history  , so ;


Solution 1 :  I can implement a chat history and chatgoback(),chatgoforward() functionality instead of using browser history while navigating to go back and forward.  

Solution 2 : I can change the current history implementation to store only the chat links.

Solution 3 : I can debug the current navigation history and leave it as is , make it work on firefox and solve the issue that while moving in history , ```
useLinking.js:33 Uncaught TypeError: Cannot read property 'id' of undefined
    at useLinking.js:33
```

**Comment by parasharrajat** [comment]
Thanks for the investigation. As this issue involves working with navigate state,  a solution should correct that to achieve the proper navigation. And it should work the same on all devices. 
1. Android, back button. 
2. Browser <- and  -> buttons regardless of browser. 

No hacks which directly work with browser History API. We use react-navigation so solution should utilize that purely for the prupose.

**Comment by murataka** [comment]
> Thanks for the investigation. As this issue involves working with navigate state, a solution should correct that to achieve the proper navigation. And it should work the same on all devices.



> ooooh, it might have to do with chat switcher, seems to work as expected when accessing chats via LHN
> 
> * click Rajat in LHN
> * click Mitch in LHN
> * click Monte in LHN
> * ⌘[ goes to Mitch
> * ⌘[ goes to Rajat.
> * ⌘] goes to Mitch
> * ⌘] goes to Monte
> * ⌘[ goes nowhere (hahahhahahah)

Ok , so is it the third one you request ? 

> debug the current navigation history and leave it as is , make it work on firefox and solve the issue that while moving in history ,

That will solve the issue across all devices.

**Comment by murataka** [comment]
> No hacks which directly work with browser History API. We use react-navigation so the solution should utilize that purely for the purpose.

currently 

`browserWindow.webContents.goForward`

is being used , and we shall change it to use the react components , am I correct  ?

**Comment by parasharrajat** [comment]
`browserWindow.webContents.goForward` is a direct API to press the Forward button on your browser. It utilizes the navigation history stack. I don't think this needs to be changed.  But happy to hear alternatives that fixes the issue.

**Comment by murataka** [comment]
> `browserWindow.webContents.goForward` is a direct API to press the Forward button

we should use react router  for historyfor both native and browser , that will make sure there it will work on all platforms .  because react already overrides that functionality in browser. That will make it run correctly on all platforms.

**Comment by murataka** [comment]
> `browserWindow.webContents.goForward` is a direct API to press the Forward button on your browser. It utilizes the navigation history stack. I don't think this needs to be changed. But happy to hear alternatives that fixe the issue.

https://stackoverflow.com/questions/42701129/how-to-push-to-history-in-react-router-v4
Creating a new browserHistory won't work because <BrowserRouter> creates its own history instance, and listens for changes on that. So a different instance will change the url but not update the <BrowserRouter>.
browserHistory is not exposed by react-router in v4, only in v2.

**Comment by parasharrajat** [comment]
Could you please do more research and gather information before we discuss a solution here? 
1. we use react-navigation. 
2. `history.forward()` === `browserWindow.webContents.goForward`.

**Comment by ahmdshrif** [comment]
# Proposal 
TL;

first, I want to explain the issue more and why I am going to this solution. 

## Issue :
this issue is not related to desktop shortcuts as (electron config ) but it's a general issue in web history.
the issue happened when you navigate from the model(like search)  to the report route this reset the history of the main route.
the history works correctly if you just navigate from the drawer. 

## Root cause : 
in code when navigating from search back to the drawer with new report id 
we dispatch custom action called (pushDrawerRoute) and it's called reset action but we already update the history.
but look like it's not working.

i notice we dispatch many actions at the same time `drawer.popTop`, `goBack` and  `reset`. 

and navigation does these actions parallel, not series. and that somehow leads to an async issue and lead to  loss of navigation history.
the problem here is navigation does not return promises so we know the first action is finished.

so to work around that I change to code (will explain this in the proposal section) set some time between `goback` action and `rese`t action.  and that solves the issue!!.

## Some Solutions : 
I know there is some suggestion to have a petter solution but I try most of them but looks like react-navigation has an issue with that. 

1- I try to reset the root state . and remove the search route and add history to the home route.  I expect this to handle the 2 actions on one action but not work fine .

2- I try to push the route with new param so we push a new route, not just params history. 
that's work but when you go pack the page reload not just update the state. 






# Final Proposal, 

in src/libs/Navigation/Navigation.js  : -
finally, this is how I think this issue can be solved. :

I will add this:- 

```ts
    if (isDrawerRoute(route)) {
        const numberOfRoutes = navigationRef.current.getState().routes.length;
        if (numberOfRoutes > 1) {
            CustomActions.navigateBackToRootDrawer(route);
        }

        // dispatch new actions after first action is executed.
        setTimeout(() => {
            navigationRef.current.dispatch(CustomActions.pushDrawerRoute(route));
        });
        return;
    }
```

and delete these lines from src/libs/Navigation/CustomActions.js 
since we handle this case above and this code is useless now or keep it in case of use it in another place in the future. 

```ts
        // When we are navigating away from a non-drawer navigator we need to first dismiss any screens pushed onto the main stack.
        if (currentState.type !== 'drawer') {
            navigateBackToRootDrawer();
        }
```


# Preview: - 



https://user-images.githubusercontent.com/21364901/151598606-bf80c24f-ca12-4399-a571-952bee656ea1.mov

**Comment by murataka** [comment]
Proposal 

in rect-navigation repo , @roryabraham already worked on that issue.

However ,
<img width="539" alt="Screen Shot 2022-01-28 at 22 48 18" src="https://user-images.githubusercontent.com/5358438/151611960-92006884-7b2a-4a4c-8f61-3efdcb4ec657.png">

As you see in screenshot , the index 1 is missing after doing some back and forward.
uselinking.tsx  108
```

        items = [{ path, state, id }];
      } else {
        items[index] = { path, state, id };
      }
 
```

is some dangerous piece of code and is suspected for the issue.

As the issue is external , I can continue to the fix @roryabraham  implemented  at react-navigation.

there might be a workaround on react-navigation like

uselinking.tsx  187 
` items.findIndex(item =>item&& item.id === id)`

,but I want to work on it to solve the issue not using any workaround.

**Comment by brentvatne** [comment]
for future reference, @satya164 is the primary maintainer of react navigation and is available for hire to work on these types of issues

**Comment by parasharrajat** [comment]
Both of you guys have interesting solutions. But if this issue lies with React navigation, I would prefer it to be fixed upstream. 
Great initiative @murataka. Thanks for creating the PR upstream. 

Have you tested the fix on our repo? Does it fix the issue here? Also, can I have a test branch to test with? It complete third party so it would make more sense.

**Comment by murataka** [comment]
> Have you tested the fix on our repo? Does it fix the issue here? Also, can I have a test branch to test with? It is completely the third party so it would make more sense.

As expected , the workaround will possibly cause issues, yes it seems to work at our repo , but I will fix it as a proper solution , and sleep comfortably.

**Comment by murataka** [comment]
> Have you tested the fix on our repo? Does it fix the issue here? Also, can I have a test branch to test with? It is completely the third party so it would make more sense.

Yes , this is the case , but I did not hesitate to fix it on that repo , with the expectation I can be helpful a bit to @satya164 the maintainer of the repo..

**Comment by murataka** [comment]
I shall add the test steps after I finish it and commit to the actual react-navigator repo.

**Comment by parasharrajat** [comment]
I will wait to hear from you. Also, please share the test branch when ready. We are not in hurry and this issue can wait.

**Comment by murataka** [comment]
> 

I need to eat expensify's pizza asap .

**Comment by parasharrajat** [comment]
I am not sure how can I test without the test branch. But if you have a way. you can tell me so that I can test it and if that works, I can get this issue assigned to you.

**Comment by murataka** [comment]
I shall have a proper solution in a few hours for sure.

**Comment by murataka** [comment]
I think it is finished, paste the below file on 

@react-navigation/lib/module/useLinking.js for testing purposes. 

I shall create a pull request , hope the maintainer likes and accepts the solution.

**Comment by murataka** [comment]
```
import { findFocusedRoute, getActionFromState as getActionFromStateDefault, getPathFromState as getPathFromStateDefault, getStateFromPath as getStateFromPathDefault } from '@react-navigation/core';
import { nanoid } from 'nanoid/non-secure';
import * as React from 'react';
import ServerContext from './ServerContext';

const createMemoryHistory = () => {
  let index = 0;
  let items = []; // Pending callbacks for `history.go(n)`
  // We might modify the callback stored if it was interrupted, so we have a ref to identify it

  const pending = [];

  const interrupt = () => {
    // If another history operation was performed we need to interrupt existing ones
    // This makes sure that calls such as `history.replace` after `history.go` don't happen
    // Since otherwise it won't be correct if something else has changed
    pending.forEach(it => {
      const cb = it.cb;

      it.cb = () => cb(true);
    });
  };

  const history = {
    get index() {
      var _window$history$state;

      // We store an id in the state instead of an index
      // Index could get out of sync with in-memory values if page reloads
      const id = (_window$history$state = window.history.state) === null || _window$history$state === void 0 ? void 0 : _window$history$state.id;

      if (id) {
        const index = items.findIndex(item => item&&item.id === id);
        return index > -1 ? index : 0;
      }

      return 0;
    },

    get(index) {
      return items[index];
    },

    backIndex({
      path
    }) {
      // We need to find the index from the element before current to get closest path to go back to
      for (let i = index - 1; i >= 0; i--) {
        const item = items[i];

        if (item.path === path) {
          return i;
        }
      }

      return -1;
    },

    push({
      path,
      state
    }) {
      interrupt();
      const id = nanoid(); // When a new entry is pushed, all the existing entries after index will be inaccessible
      // So we remove any existing entries after the current index to clean them up

      items = items.slice(0, index + 1);
      items.push({
        path,
        state,
        id
      });
      index = items.length - 1; // We pass empty string for title because it's ignored in all browsers except safari
      // We don't store state object in history.state because:
      // - browsers have limits on how big it can be, and we don't control the size
      // - while not recommended, there could be non-serializable data in state

      window.history.pushState({
        id
      }, '', path);
    },

    replace({
      path,
      state
    }) {
      var _window$history$state2, _window$history$state3;

      interrupt();
      const id = (_window$history$state2 = (_window$history$state3 = window.history.state) === null || _window$history$state3 === void 0 ? void 0 : _window$history$state3.id) !== null && _window$history$state2 !== void 0 ? _window$history$state2 : nanoid();

      if (!items.length || items.findIndex(item =>item&& item.id === id) < 0) {
        // There are two scenarios for creating an array with only one history record:
        // - When loaded id not found in the items array, this function by default will replace
        //   the first item. We need to keep only the new updated object, otherwise it will break
        //   the page when navigating forward in history.
        // - This is the first time any state modifications are done
        //   So we need to push the entry as there's nothing to replace
        items = [{
          path,
          state,
          id
        }];
      } else {
        items [0]={
          path,
          state,
          id
        };
      }

      window.history.replaceState({
        id
      }, '', path);
    },

    // `history.go(n)` is asynchronous, there are couple of things to keep in mind:
    // - it won't do anything if we can't go `n` steps, the `popstate` event won't fire.
    // - each `history.go(n)` call will trigger a separate `popstate` event with correct location.
    // - the `popstate` event fires before the next frame after calling `history.go(n)`.
    // This method differs from `history.go(n)` in the sense that it'll go back as many steps it can.
    go(n) {
      interrupt();

      if (n > 0) {
        // We shouldn't go forward more than available index
        n = Math.min(n, items.length - 1);
      } else if (n < 0) {
        // We shouldn't go back more than the 0 index
        // Otherwise we'll exit the page
        n = index + n < 0 ? -index : n;
      }

      if (n === 0) {
        return;
      }

      index += n; // When we call `history.go`, `popstate` will fire when there's history to go back to
      // So we need to somehow handle following cases:
      // - There's history to go back, `history.go` is called, and `popstate` fires
      // - `history.go` is called multiple times, we need to resolve on respective `popstate`
      // - No history to go back, but `history.go` was called, browser has no API to detect it

      return new Promise((resolve, reject) => {
        const done = interrupted => {
          clearTimeout(timer);

          if (interrupted) {
            reject(new Error('History was changed during navigation.'));
            return;
          } // There seems to be a bug in Chrome regarding updating the title
          // If we set a title just before calling `history.go`, the title gets lost
          // However the value of `document.title` is still what we set it to
          // It's just not displayed in the tab bar
          // To update the tab bar, we need to reset the title to something else first (e.g. '')
          // And set the title to what it was before so it gets applied
          // It won't work without setting it to empty string coz otherwise title isn't changing
          // Which means that the browser won't do anything after setting the title


          const {
            title
          } = window.document;
          window.document.title = '';
          window.document.title = title;
          resolve();
        };

        pending.push({
          ref: done,
          cb: done
        }); // If navigation didn't happen within 100ms, assume that it won't happen
        // This may not be accurate, but hopefully it won't take so much time
        // In Chrome, navigation seems to happen instantly in next microtask
        // But on Firefox, it seems to take much longer, around 50ms from our testing
        // We're using a hacky timeout since there doesn't seem to be way to know for sure

        const timer = setTimeout(() => {
          const index = pending.findIndex(it => it.ref === done);

          if (index > -1) {
            pending[index].cb();
            pending.splice(index, 1);
          }
        }, 100);

        const onPopState = () => {
          var _window$history$state4;

          const id = (_window$history$state4 = window.history.state) === null || _window$history$state4 === void 0 ? void 0 : _window$history$state4.id;
          const currentIndex = items.findIndex(item =>item&& item.id === id); // Fix createMemoryHistory.index variable's value
          // as it may go out of sync when navigating in the browser.

          index = Math.max(currentIndex, 0);
          const last = pending.pop();
          window.removeEventListener('popstate', onPopState);
          last === null || last === void 0 ? void 0 : last.cb();
        };

        window.addEventListener('popstate', onPopState);
        window.history.go(n);
      });
    },

    // The `popstate` event is triggered when history changes, except `pushState` and `replaceState`
    // If we call `history.go(n)` ourselves, we don't want it to trigger the listener
    // Here we normalize it so that only external changes (e.g. user pressing back/forward) trigger the listener
    listen(listener) {
      const onPopState = () => {
        if (pending.length) {
          // This was triggered by `history.go(n)`, we shouldn't call the listener
          return;
        }

        listener();
      };

      window.addEventListener('popstate', onPopState);
      return () => window.removeEventListener('popstate', onPopState);
    }

  };
  return history;
};
/**
 * Find the matching navigation state that changed between 2 navigation states
 * e.g.: a -> b -> c -> d and a -> b -> c -> e -> f, if history in b changed, b is the matching state
 */


const findMatchingState = (a, b) => {
  if (a === undefined || b === undefined || a.key !== b.key) {
    return [undefined, undefined];
  } // Tab and drawer will have `history` property, but stack will have history in `routes`


  const aHistoryLength = a.history ? a.history.length : a.routes.length;
  const bHistoryLength = b.history ? b.history.length : b.routes.length;
  const aRoute = a.routes[a.index];
  const bRoute = b.routes[b.index];
  const aChildState = aRoute.state;
  const bChildState = bRoute.state; // Stop here if this is the state object that changed:
  // - history length is different
  // - focused routes are different
  // - one of them doesn't have child state
  // - child state keys are different

  if (aHistoryLength !== bHistoryLength || aRoute.key !== bRoute.key || aChildState === undefined || bChildState === undefined || aChildState.key !== bChildState.key) {
    return [a, b];
  }

  return findMatchingState(aChildState, bChildState);
};
/**
 * Run async function in series as it's called.
 */


const series = cb => {
  // Whether we're currently handling a callback
  let handling = false;
  let queue = [];

  const callback = async () => {
    try {
      if (handling) {
        // If we're currently handling a previous event, wait before handling this one
        // Add the callback to the beginning of the queue
        queue.unshift(callback);
        return;
      }

      handling = true;
      await cb();
    } finally {
      handling = false;

      if (queue.length) {
        // If we have queued items, handle the last one
        const last = queue.pop();
        last === null || last === void 0 ? void 0 : last();
      }
    }
  };

  return callback;
};

let linkingHandlers = [];
export default function useLinking(ref, {
  independent,
  enabled = true,
  config,
  getStateFromPath = getStateFromPathDefault,
  getPathFromState = getPathFromStateDefault,
  getActionFromState = getActionFromStateDefault
}) {
  React.useEffect(() => {
    if (process.env.NODE_ENV === 'production') {
      return undefined;
    }

    if (independent) {
      return undefined;
    }

    if (enabled !== false && linkingHandlers.length) {
      console.error(['Looks like you have configured linking in multiple places. This is likely an error since deep links should only be handled in one place to avoid conflicts. Make sure that:', "- You don't have multiple NavigationContainers in the app each with 'linking' enabled", '- Only a single instance of the root component is rendered'].join('\n').trim());
    }

    const handler = Symbol();

    if (enabled !== false) {
      linkingHandlers.push(handler);
    }

    return () => {
      const index = linkingHandlers.indexOf(handler);

      if (index > -1) {
        linkingHandlers.splice(index, 1);
      }
    };
  }, [enabled, independent]);
  const [history] = React.useState(createMemoryHistory); // We store these options in ref to avoid re-creating getInitialState and re-subscribing listeners
  // This lets user avoid wrapping the items in `React.useCallback` or `React.useMemo`
  // Not re-creating `getInitialState` is important coz it makes it easier for the user to use in an effect

  const enabledRef = React.useRef(enabled);
  const configRef = React.useRef(config);
  const getStateFromPathRef = React.useRef(getStateFromPath);
  const getPathFromStateRef = React.useRef(getPathFromState);
  const getActionFromStateRef = React.useRef(getActionFromState);
  React.useEffect(() => {
    enabledRef.current = enabled;
    configRef.current = config;
    getStateFromPathRef.current = getStateFromPath;
    getPathFromStateRef.current = getPathFromState;
    getActionFromStateRef.current = getActionFromState;
  });
  const server = React.useContext(ServerContext);
  const getInitialState = React.useCallback(() => {
    let value;

    if (enabledRef.current) {
      var _server$location;

      const location = (_server$location = server === null || server === void 0 ? void 0 : server.location) !== null && _server$location !== void 0 ? _server$location : typeof window !== 'undefined' ? window.location : undefined;
      const path = location ? location.pathname + location.search : undefined;

      if (path) {
        value = getStateFromPathRef.current(path, configRef.current);
      }
    }

    const thenable = {
      then(onfulfilled) {
        return Promise.resolve(onfulfilled ? onfulfilled(value) : value);
      },

      catch() {
        return thenable;
      }

    };
    return thenable; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const previousIndexRef = React.useRef(undefined);
  const previousStateRef = React.useRef(undefined);
  const pendingPopStatePathRef = React.useRef(undefined);
  React.useEffect(() => {
    previousIndexRef.current = history.index;
    return history.listen(() => {
      var _previousIndexRef$cur;

      const navigation = ref.current;

      if (!navigation || !enabled) {
        return;
      }

      const path = location.pathname + location.search;
      const index = history.index;
      const previousIndex = (_previousIndexRef$cur = previousIndexRef.current) !== null && _previousIndexRef$cur !== void 0 ? _previousIndexRef$cur : 0;
      previousIndexRef.current = index;
      pendingPopStatePathRef.current = path; // When browser back/forward is clicked, we first need to check if state object for this index exists
      // If it does we'll reset to that state object
      // Otherwise, we'll handle it like a regular deep link

      const record = history.get(index);

      if ((record === null || record === void 0 ? void 0 : record.path) === path && record !== null && record !== void 0 && record.state) {
        navigation.resetRoot(record.state);
        return;
      }

      const state = getStateFromPathRef.current(path, configRef.current); // We should only dispatch an action when going forward
      // Otherwise the action will likely add items to history, which would mess things up

      if (state) {
        // Make sure that the routes in the state exist in the root navigator
        // Otherwise there's an error in the linking configuration
        const rootState = navigation.getRootState();

        if (state.routes.some(r => !(rootState !== null && rootState !== void 0 && rootState.routeNames.includes(r.name)))) {
          console.warn("The navigation state parsed from the URL contains routes not present in the root navigator. This usually means that the linking configuration doesn't match the navigation structure. See https://reactnavigation.org/docs/configuring-links for more details on how to specify a linking configuration.");
          return;
        }

        if (index > previousIndex) {
          const action = getActionFromStateRef.current(state, configRef.current);

          if (action !== undefined) {
            try {
              navigation.dispatch(action);
            } catch (e) {
              // Ignore any errors from deep linking.
              // This could happen in case of malformed links, navigation object not being initialized etc.
              console.warn(`An error occurred when trying to handle the link '${path}': ${e.message}`);
            }
          } else {
            navigation.resetRoot(state);
          }
        } else {
          navigation.resetRoot(state);
        }
      } else {
        // if current path didn't return any state, we should revert to initial state
        navigation.resetRoot(state);
      }
    });
  }, [enabled, history, ref]);
  React.useEffect(() => {
    var _ref$current;

    if (!enabled) {
      return;
    }

    if (ref.current) {
      // We need to record the current metadata on the first render if they aren't set
      // This will allow the initial state to be in the history entry
      const state = ref.current.getRootState();

      if (state) {
        var _route$path;

        const route = findFocusedRoute(state);
        const path = (_route$path = route === null || route === void 0 ? void 0 : route.path) !== null && _route$path !== void 0 ? _route$path : getPathFromStateRef.current(state, configRef.current);

        if (previousStateRef.current === undefined) {
          previousStateRef.current = state;
        }

        history.replace({
          path,
          state
        });
      }
    }

    const onStateChange = async () => {
      var _route$path2;

      const navigation = ref.current;

      if (!navigation || !enabled) {
        return;
      }

      const previousState = previousStateRef.current;
      const state = navigation.getRootState();
      const pendingPath = pendingPopStatePathRef.current;
      const route = findFocusedRoute(state);
      const path = (_route$path2 = route === null || route === void 0 ? void 0 : route.path) !== null && _route$path2 !== void 0 ? _route$path2 : getPathFromStateRef.current(state, configRef.current);
      previousStateRef.current = state;
      pendingPopStatePathRef.current = undefined; // To detect the kind of state change, we need to:
      // - Find the common focused navigation state in previous and current state
      // - If only the route keys changed, compare history/routes.length to check if we go back/forward/replace
      // - If no common focused navigation state found, it's a replace

      const [previousFocusedState, focusedState] = findMatchingState(previousState, state);

      if (previousFocusedState && focusedState && // We should only handle push/pop if path changed from what was in last `popstate`
      // Otherwise it's likely a change triggered by `popstate`
      path !== pendingPath) {
        const historyDelta = (focusedState.history ? focusedState.history.length : focusedState.routes.length) - (previousFocusedState.history ? previousFocusedState.history.length : previousFocusedState.routes.length);

        if (historyDelta > 0) {
          // If history length is increased, we should pushState
          // Note that path might not actually change here, for example, drawer open should pushState
          history.push({
            path,
            state
          });
        } else if (historyDelta < 0) {
          // If history length is decreased, i.e. entries were removed, we want to go back
          const nextIndex = history.backIndex({
            path
          });
          const currentIndex = history.index;

          try {
            if (nextIndex !== -1 && nextIndex < currentIndex) {
              // An existing entry for this path exists and it's less than current index, go back to that
              await history.go(nextIndex - currentIndex);
            } else {
              // We couldn't find an existing entry to go back to, so we'll go back by the delta
              // This won't be correct if multiple routes were pushed in one go before
              // Usually this shouldn't happen and this is a fallback for that
              await history.go(historyDelta);
            } // Store the updated state as well as fix the path if incorrect


            history.replace({
              path,
              state
            });
          } catch (e) {// The navigation was interrupted
          }
        } else {
          // If history length is unchanged, we want to replaceState
          history.replace({
            path,
            state
          });
        }
      } else {
        // If no common navigation state was found, assume it's a replace
        // This would happen if the user did a reset/conditionally changed navigators
        history.replace({
          path,
          state
        });
      }
    }; // We debounce onStateChange coz we don't want multiple state changes to be handled at one time
    // This could happen since `history.go(n)` is asynchronous
    // If `pushState` or `replaceState` were called before `history.go(n)` completes, it'll mess stuff up


    return (_ref$current = ref.current) === null || _ref$current === void 0 ? void 0 : _ref$current.addListener('state', series(onStateChange));
  });
  return {
    getInitialState
  };
}
//# sourceMappingURL=useLinking.js.map

```

**Comment by murataka** [comment]
https://github.com/react-navigation/react-navigation/pull/10306

**Comment by murataka** [comment]
@parasharrajat , I shall continue to testing, there might be more fixes needed, don't know if need to fix all bugs on the library about navigation history, nor I shall be able to. let me know about your test.

**Comment by ahmdshrif** [comment]
@parasharrajat  I think there 2 cases that lead to history  issues 

1- reset the navigation when back from the search screen and that is what  @mallenexpensify explains in this comment  https://github.com/Expensify/App/issues/4612#issuecomment-906840431 and I solve it on my proposal. 

2- navigation loss when you go back and forward a lot of time and @murataka  doing amazing work with solving that from react-navigation. but I test his solution did not work for the first case.

**Comment by K4tsuki** [comment]
@ahmdshrif I think your [proposal](https://github.com/Expensify/App/issues/4612#issuecomment-1024489080) is similar to my [proposal](https://github.com/Expensify/App/issues/6912#issuecomment-1004244650)
I did mention `setTimeout` and pull out `navigateBackToRootDrawer` there

**Comment by ahmdshrif** [comment]
@K4tsuki iam not see your proposal before but in this issue will not work without timeout maybe work for the secend one .

**Comment by murataka** [comment]
> 2- navigation loss history when you go back and forward many times and @murataka doing amazing work with solving that from react-navigation. but I test his solution did not work for the first case.

As this is an external library, and all the core problem is already appearing from that particular external bug. 
I would like to see what cases you get it not work properly, but It would help much if you share some video recording, what error you get , and when it happens, and you should better do comment on react-navigation/react-navigation#10306 . 

I'm pretty sure this what I proposed was already is the only acceptable solution, but in case you want to share information about my solution , I would like some more technical notes as well ( on react-navigation/react-navigation#10306  ), not here as much as possible for this issue.

**Comment by murataka** [comment]
A note to who want to test : 

As copying and pasting the file is not the proper way of testing it, and while doing the back and forward uses caching when possible , be sure to repeat the test on maybe on an other browser or  be sure to clear app cache. 
Thanks for anyone testing it also , as soon as you share any information about the bug . But as suggested above , there seems to be no urgency. I am just trying to be helpful, and don't think have to solve the bug at all , but prove of the core problem was possibly already enough for the proposal.  I hope all is clear , and yes I shall continue to work on it as much as I want . Agains thanks for any additional information .

**Comment by ahmdshrif** [comment]
thanks for sharing that.
@parasharrajat  will test this so no need to retest from my side .
I just think there is 2 issue and we need both pr to solve them .  that is what I mean in my comment. 
if your proposal solves the 2 cases I mention it will be the only accepted one I agree with this.

**Comment by murataka** [comment]
If you don't get a result like that ( both native or browser ) , you may not be testing it correctly . I don't know if I can be able to prepare a more easy way to test this , open for suggestion. 
https://user-images.githubusercontent.com/5358438/151664851-a27dc533-a97d-4ec3-9ea2-34812bfa6573.mov

**Comment by murataka** [comment]
An other test ...

https://user-images.githubusercontent.com/5358438/151666639-c2765ffb-16e8-4bcd-b5d3-de0e6a5cc7ef.mov

**Comment by ahmdshrif** [comment]
I mean Don't navigate from drawer . 
Navigate from search page . That will reset the history . That's the case not work .

**Comment by murataka** [comment]
> I mean Don't navigate from drawer .
> Navigate from search page . That will reset the history . That's the case not work .

Seems to work correctly again , be sure to make node_modules/@react-navigation/lib/module/useLinking.js file copy paste  on the latest main branch , if you get any error : _`id of undefined`_  errors , still you may not be using the new changes.

**Comment by murataka** [comment]
> Navigate from search page . That will reset the history . That's the case not work .

I don't know if this is a bug, the problem with that is , the react-navigation library might already supposed to reset it ( for now). And in a later version , someone may request to change it. But seems to be a case for the external library, nothing seems wrong with current extensify about navigation.

**Comment by murataka** [comment]
@ahmdshrif , the navigation history appears under search history, although I did not search , but just navigated. Maybe this is causing some confusion ?

**Comment by ahmdshrif** [comment]
My be i have some cache as you mentioned . But i recommend you do the same steps in the vedio in here https://github.com/Expensify/App/issues/4612#issuecomment-1024489080

Jsut search for many pepopl 
And navigate back if it back to correct user . Its fine .

**Comment by murataka** [comment]
> But i recommend you do the same steps in the vedio in here [#4612 (comment)](https://github.com/Expensify/App/issues/4612#issuecomment-1024489080)

The problem with that is , we have no search history.  Seems as if someone has used the navigation history as search history 
 or vice versa, 

The question is , **_do we have a search history_** ? That seems to be an other bug. Search history uses navigation history. That is what I understood is happening.

The issue related to this is "The Back ⌘[ and Forward ⌘] shortcuts keys are not working as expected" . I am not the one to review proposals :D so you can discuss if it is related or not . This is my opinion of course .

**Comment by ahmdshrif** [comment]
i work in search senario because of this comment from reporter https://github.com/Expensify/App/issues/4612#issuecomment-906840431

Note cmd+k is the search shortcut .

**Comment by murataka** [comment]
> i work in search senario because of this comment from reporter [#4612 (comment)](https://github.com/Expensify/App/issues/4612#issuecomment-906840431)
> 
> Note cmd+k is the search shortcut .



> ## Action Performed:
> 1. Open desktop app
> 2. Navigate to several conversations
> 3. Use the back `⌘[` and Forward `⌘]` shortcuts
> 
> ## Expected Result:
> Back `⌘[` and Forward `⌘]` shortcuts should work and navigate through the previously navigated conversations
> 
> ## Actual Result:
> Shortcuts are not working as expected.

And #4612 (comment) is related to not having a search history implemented. We also discussed above if implementing an other history object would be fine or not , but @parasharrajat  did not like that proposal. 

However , this seems to be an other bug ( We need a seperate search history ) .

**Comment by murataka** [comment]
@ahmdshrif , if you think it is related to this subject , I shall go open a bug report on slack. If you think this is not related what you propose , you should report it on slack channel I think .

**Comment by ahmdshrif** [comment]
I think  i solve the main issue accourding to the comments i mentioned. So we wait for @parasharrajat  and @NikkiWines to validat wich issue is the main  and wich is the related one .  Or both is will be the  main issue .

**Comment by parasharrajat** [comment]
Hohoho, there is some heavy discussion happening here :wink: We will get to every issue one by one. but I am having hard time understanding the whole discussion above. 

Can someone please summarize both issues? I will try to reproduce both and see. 

--- 

@murataka Sorry, I didn't find any file at `node_modules/@react-navigation/lib/module/useLinking.js` path. The structure goes like this. What is the main module `core`?
![image](https://user-images.githubusercontent.com/24370807/151694348-13d016b6-2b96-4cf0-904f-3f45270aa8a0.png)

**Comment by murataka** [comment]
![here](https://user-images.githubusercontent.com/5358438/151694459-d7830c79-c50f-4d38-a507-a4290019b9c0.png)

**Comment by murataka** [comment]
You better do a file sarch for useLinking.js

in node_modules folder.

**Comment by parasharrajat** [comment]
To clarify the scope of this
1. Report navigation order should be maintained on all cases. Whether the report is opened from LHN(left hand navigation) or RHN(right hand navigation). 
2. Also, when we open search page and open a report. navigating back should take you to previous report not the search page. 

There are many navigational issues #7363 . It is possible that solution for one can solve others. This is known to us and must have been mentioned before. In that case we will discontinue other which is automatically solved. This is one of the reason, most of them are still open. We are looking for a concrete solution. 

@murataka I tested your solution. 

LHN report navigation seems to work better now. But navigation is not remembered for navigating from search page. 

Let's try fix that as well. 
--- 

For general discussion, please keep it off the issue thread so that readers are not confused.

**Comment by murataka** [comment]
> LHN report navigation seems to work better now. But navigation is not remembered for navigating from search page.

It displays the navigation history , and this is an other issue . In fact , Search history should display search history instead of navigation history.

That creates reset-like behavior on navigation history.

**Comment by parasharrajat** [comment]
When we open a report from the search page, it navigates to that report and `⌘[  and ⌘]` just go back and forward in the browser navigation history. When proper navigation history will be maintained, they will just start to work. 

I don't think they are separate issues. This issue was originally faced from [these steps](https://github.com/Expensify/App/issues/4612#issuecomment-906840431). 

If you think they are separate, you can raise a slack thread. Others can help in deciding that.

**Comment by murataka** [comment]
> If you think they are separate, you can raise a slack thread. Others can help in deciding that.

I don't think I'm the one to have a conclusion on that , please check the comments in CustomActions.js .

**Comment by parasharrajat** [comment]
Correct, we can just focus on the reproduction steps [here](https://github.com/Expensify/App/issues/4612#issuecomment-906840431). This should fix that. All of the other issues have separate tasks/GH issues.

**Comment by parasharrajat** [comment]
@ahmdshrif Thanks for the proposal. I will review that as well.

**Comment by ahmdshrif** [comment]
thanks @parasharrajat waiting for your feedback.

**Comment by murataka** [comment]
> // Note: A current limitation with this is that navigating "back" won't display the routes we have cleared out e.g. SearchPage and the history effectively gets "reset".

@parasharrajat  I think that you have some kind of distraction issues like me, in case you missed it, that is the root cause of 
reset-like behavior mentioned above.

**Comment by murataka** [comment]
Proposal ( this is solving the RHN related reset issue. Any bonus is appreciated )

In CustomActions.js

line 129 ..
```
        const routes = state.routes;

        return CommonActions.reset({
            ...state,
            routes: [{
                name: newScreenName,
                params: newScreenParams,
            }],
            index: routes.length - 1,
            history,
        });
```

solves this issue.

**Comment by ahmdshrif** [comment]
since I spend many days working on changing the state in reset action, I can confirm that adding an index will not solve the RHN issue.
but you can make a test also.

**Comment by murataka** [comment]
> 

Without the react-navigation library fix, I totally understand you spending so much time on that. But that works after combining those fixes.

**Comment by murataka** [comment]
@parasharrajat , @ahmdshrif had some help on me focusing on the search (RHN) related issues. If possible make sure he also gets hired if my fixes are accepted.

**Comment by ahmdshrif** [comment]
I just remember I try this fix and not work.  if it worked with you that's fine. it's better than my suggestion.

**Comment by ahmdshrif** [comment]
thanks, @murataka , if your fix is working it's your job bro ❤️
I am not suggesting that.

**Comment by mallenexpensify** [comment]
Doubled price to $8000 https://www.upwork.com/jobs/~01a4e8881ff6937bae

**Comment by murataka** [comment]
@parasharrajat , 
reactnative/navigation library has the proper solution merged now . but 

package.json is still 

`    "@react-navigation/native": "6.0.6",`

and should be upgraded to

`    "@react-navigation/native": "6.0.7",`

at least.

This has many impacts on other navigation related issues (and also some issues which seems to be unrelated ). I think update should be done asap , to prevent further related bug reports about navigation to getting whitescreen  .

**Comment by murataka** [comment]
I shall create the pr , let me know if you also want me to include RHN solution to that pr.

**Comment by parasharrajat** [comment]
Thanks to @murataka, I am sure we can compensate for your efforts but I don't consider [that](https://github.com/Expensify/App/issues/4612#issuecomment-1026657013) a complete solution to this issue. 

Let me see what is the best thing to do.  I will test your RHN solution and if that works it's great otherwise, we will find some way.

Please do not create a PR until requested to do so.

**Comment by murataka** [comment]
> Let me see what is the best thing to do. I will test your RHN solution and if that works it's great. We can assign this issue.

The RHN solution will work only if you combine the upgrading the navigation library.

**Comment by murataka** [comment]
> I hope you would find my suggestion helpful. This is not a pattern rather my thoughts so do best at your discretion.

Yes, I'm trying to get used to that . Thanks for that , I already asked that on slack before , and at last I have that detailed explanation. I shall read it again and again for sure !

**Comment by murataka** [comment]
check this one. https://github.com/Expensify/App/issues/4612#issuecomment-1025139008

**Comment by murataka** [comment]
I rechecked that , and the navigation fix still not existing in tagged versions. #4612 (comment) is fine ,the merged pr may not exist in tagged versions . unless we can use the main branch from react-native , #4612 (comment) can not be tested properly. I'm not sure if those commits affect the tagged builds. that is the fix related. ( sorry for adding more comments) https://github.com/lostintime/react-navigation/blob/1ba272fff456dfa2160c4992585fbe0fee7cf372/packages/native/src/useLinking.tsx#L110

**Comment by ahmdshrif** [comment]
# Propasal :
I explain here https://github.com/Expensify/App/issues/4612#issuecomment-1024489080 the issue of dispatch many actions at the same time. and that makes many navigation issue,s not just this.

I here have a better proposal which complains the actions on one action.  instead of workaround with time out. 

src/libs/Navigation/CustomActions.js

```ts
function pushDrawerRoute(route) {
    const rootState = getActiveState();

    // Parse the state, name, and params from the new route we want to navigate to.
    const newStateFromRoute = getStateFromPath(route, linkingConfig.config);
    const newScreenName = getScreenNameFromState(newStateFromRoute);
    const newScreenParams = getParamsFromState(newStateFromRoute);

    const state = {
        ...rootState.routes[0].state,
        index: 0,
        routes: [{
            name: newScreenName,
            params: newScreenParams,
        }],
    };

    const screenRoute = {type: 'route', name: newScreenName};
    const history = _.map(state.history ? [...state.history] : [screenRoute], () => screenRoute);

    // Force drawer to close and show
    history.push({
        type: 'drawer',
        status: 'closed',
    });
    state.history = history;

    if (rootState.routes.length > 1) {
        const homRoute = {
            ...rootState.routes[0],
            state,
        };

        const action = getActionFromState({
            index: 0,
            routes: [homRoute],
        });
        navigationRef.current.dispatch(action);
    } else {
        navigationRef.current.dispatch(CommonActions.reset(state));
    }
}
```


src/libs/Navigation/Navigation.js

```ts
if (isDrawerRoute(route)) {
        CustomActions.pushDrawerRoute(route);
        return;
    }
```


this proposal is built on top of fix form @murataka  on react-navigation. 
so we need both pr but that will fix the whole issues.

**Comment by parasharrajat** [comment]
I am tackling the navigation issues one by one as it is possible one indirectly solves the other. Thanks for your patience.

**Comment by murataka** [comment]
> I am tackling the navigation issues one by one as it is possible one indirectly solves the other. Thanks for your patience.

Just share it if we can be helpful on any part of your review. I would offer you to use the slack more actively while reviewing proposals, and dedicate to what you are currently doing. Sure you need to multitask for some reasons.

**Comment by murataka** [comment]
The external bug is at last been included in

`"@react-navigation/native": "6.0.8",`

. I think that bug solved  has many affect if not the root cause. It strongly suggest any reviewers to update package.json while reviewing any navigational issue, as this is external , and is a potential to cause false negatives .

**Comment by roryabraham** [comment]
Of course they publish a release 1 day after we create our own fork 🙄

**Comment by murataka** [comment]
after 

`"@react-navigation/native": "6.0.8",`

is merged by @roryabraham , The Back ⌘[ and Forward ⌘] shortcuts keys seems to be working as expected , also the RHN seems to work much better now . @parasharrajat  , can you test it and share what you think ? I may need to update my proposal in case you think any further action to be taken.

**Comment by parasharrajat** [comment]
Sure, I can do that.

**Comment by murataka** [comment]
Proposal

Previously, 

I further examined the app , and we understood that , the root cause was the external navigation library is not working properly.

https://user-images.githubusercontent.com/5358438/151611960-92006884-7b2a-4a4c-8f61-3efdcb4ec657.png

Also , the root cause was definitely related to this library, and the bug was reproducible with the same error logs which we got from native. 

After further examining , we managed to have 

"@react-navigation/native": “6.0.8",

Included the bug fixes at PR merge https://github.com/Expensify/App/pull/7539 , which solves the navigation - history related problems partially .

Currently,

After further examining the expected behavior and what we get while using the navigation functions ,

We have duplicate items in the items list , and also the items list still gets lost ( reset ) after doing back and forward. 

Expected result was items should contain 6 items after back and forward. I navigated to 6 reports , then back->forward->back-forward. ( See image below )

![navigationlogofitems](https://user-images.githubusercontent.com/5358438/152494210-c8bb7b36-a0b5-4540-83b5-3deca1888c14.png)

Additionally , my previous proposal about this issue

https://github.com/Expensify/App/issues/4612#issuecomment-1025139008

Now seems to be a workaround , because when we properly have the bug described above ( @react-navigation/native related indexes bug) , that shall be non essential. 

What I propose : 

I will open an issue on @react-navigation/native library , share what the problem is , explain the bug more in more details, and start fixing it the best way possible . Currently , there are some open issues not pointing the root problem precisely, but the consequences of it. 

Chances are , I will be able to fix it soon , as I am now more familiar about the bug and the expected behavior, or someone else will be able solve the bug in a more appropriate way than it is currently implemented. That shall make we have the proper fix asap. 

Notes: My conclusions are this library is a core part of our app , and when it has issues, this causes more than one problems at our end , and as we are mostly focusing the code implemented at this repo , this error causes us ( at least me ) to go for workarounds , which should be prevented as much as possible.

**Comment by murataka** [comment]
@marcaaron , I think you have an Idea about the above proposal, after reading https://github.com/Expensify/App/issues/6912#issuecomment-1005168037 . When you have time ,I think it would help much on navigation related issues if share what you think .  Thank you !

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kevinksullivan (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by mallenexpensify** [comment]
@parasharrajat what's status here? Does @murataka's proposal need reviewing? https://github.com/Expensify/App/issues/4612#issuecomment-1029745730

@kevinksullivan  I'm out next week so auto-added you to keep 👀 on this til I'm back, then I'll take back over.

**Comment by parasharrajat** [comment]
@mallenexpensify I was busy with some other navigation issue and now I will pick the second on the list. This is third on that list. So sadly, we have to wait a bit on this but I will take look on the latest proposal next week.

**Comment by ahmdshrif** [comment]
My proposal also need reviewing 😃

**Comment by parasharrajat** [comment]
No update. Still holding... Made progress on the other issue. PR ready. The issue list can be found [here](https://github.com/Expensify/App/issues/7363).

**Comment by NikkiWines** [comment]
@parasharrajat what are we holding on? Are we tackling the issues listed [here](https://github.com/Expensify/App/issues/7363) in order? If so, we should add `[HOLD]` to the issues (like this one) not yet being worked on.

**Comment by parasharrajat** [comment]
Yeah, I am handling the issues in order (the order is not final). But the main reason is to fix issues one by one. Solutions can overlap. But I am also tracking proposals on each issue. 

Yes, you can add `[hold]` if needed. I don't have access to do that :sweat_smile: .

**Comment by NikkiWines** [comment]
Cool, adding the HOLD and dropping this to a weekly then

**Comment by mallenexpensify** [comment]
@parasharrajat will get to soon

**Comment by parasharrajat** [comment]
Update: This one is close. This is next on the list. I have already reviewed the initial proposals But I need to understand the new suggested changes.

**Comment by murataka** [comment]
@parasharrajat , I think it should be helpful to create a discussion on slack , for at least me , to clarify   the last changes and what they are doing , especially at customactions.js , the dispach calls. Also it would be helpful if anyone enlighten me about what they are really doing .  I am unable to understand those calls deeply, but just trying to guess , but I see that you are beyond me on that. Thanks !

**Comment by parasharrajat** [comment]
I am trying to find solutions for other issues on that list. You can see the updates on those issues. If I find something that needs attention, I will open a discussion on slack and involve you. 

Sure, I will be happy to help you with the CustomActions based on your proposed changes.

**Comment by murataka** [comment]
the complete solution is at slack channel 

https://expensify.slack.com/archives/C01GTK53T8Q/p1645912640776129?thread_ts=1645912640.776129&cid=C01GTK53T8Q

**Comment by parasharrajat** [comment]
This is no more on hold. Waiting for proposals... and reviewing old ones.

I appreciate your proposal @murataka and I will take a look on Slack but the correct process is to post the proposal on Github.

**Comment by NikkiWines** [comment]
Not overdue, just waiting on proposals

**Comment by ahmdshrif** [comment]
Mmm .. what do you think about my proposal here 

https://github.com/Expensify/App/issues/4612#issuecomment-1026774299

**Comment by parasharrajat** [comment]
My bad. Forgot to leave a review for your proposal @ahmdshrif. Do you mind writing a single post explaining your approach to the problem? I saw your code but it is not clear enough. 

1. What do you think the problem is? Are you clear about the expected behavior here? 
2. What is causing this issue? And how it is doing that?
3. Break down the issues into steps and try to solve each separately instead of giving a diff or code.
4. List down all the navigation scenarios that are touched by the piece of code you are planning to change. 
5. Explain how your solution will work for each of those navigation scenarios.

**Comment by mallenexpensify** [comment]
Discussing [internally](https://expensify.slack.com/archives/C02NK2DQWUX/p1648063112665829) to see how this issue relates to other navigation issues @parasharrajat has gathered in [this tracking issue. ](https://github.com/Expensify/App/issues/7363)

**Comment by mallenexpensify** [comment]
Doubled price to $16000
Needed to create a new job 
https://www.upwork.com/jobs/~01294fe6c7ab1639a8

**Comment by rushatgabhane** [comment]
> * click Rajat in LHN
     * click Mitch in LHN
     * click Monte in LHN
     * ⌘[ goes to Mitch
     * ⌘[ goes to Rajat.

@parasharrajat This is no longer true. 

This is the behavior on staging and production. 
* click Rajat in LHN
 * click Mitch in LHN
 * click Monte in LHN
 * ⌘[ goes to Rajat.

We're skipping every other user. It used to work before, have we introduced some changes which broke this?


https://user-images.githubusercontent.com/29673073/159834008-5666c747-ec0d-47d0-9882-c11eba73bcac.mov

**Comment by parasharrajat** [comment]
Thanks for sharing. I will run a few tests on weekends to observe the current state of the issue.

**Comment by rushatgabhane** [comment]
1. Open Nikki's chat from LHN
2. Open Mitch's chat from LHN
3. Cmd + K, then click Rajat
4. Cmd + K, then click Matt
5. Cmd + K, then click Rushat
6. Go back

Expected result
We're on Matt's report.

Actual result
We're on Nikki's report. i.e. Using RHN (Cmd + K) takes us `N - 1` back in history! 

<hr/>

From https://reactnavigation.org/docs/5.x/navigation-state/
> history - A list of visited items. This is an optional property and not present in all navigators. For example, it's only present in tab and drawer navigators in the core. 


LHN is a **tab navigator**, and reactnavigation keeps history for us. (though buggy)

While RHN (Cmd + K) is a stack navigator, so history isn't being maintained at all!! Isn't this the root cause? _ If so, shouldn't we fix it in reactnavigation itself?_

**Comment by rushatgabhane** [comment]
I'd suggest that we be break this issue into two seperate issues. (_**8k each**_)

- Keep this issue to solve RHN bugs.
- Create a new issue to solve the LHN bug. i.e add an optional history prop to stack navigator.

Thoughts?

**Comment by murataka** [comment]
@rushatgabhane , my fix solves this issue from the root at react navigation. let me know if you need any further explanation about that.

**Comment by NikkiWines** [comment]
not overdue, still waiting on proposals

**Comment by mallenexpensify** [comment]
OK... I got this split in two! This issue is specific to LHN and the new one https://github.com/Expensify/App/issues/8657 is specific to navigation in the main chat view/window.

Repriced this to $8000, other is $8000 too.
New job https://www.upwork.com/jobs/~01fe76a840ceb332a4

**Comment by murataka** [comment]
> 

I think the expected behavior should be explained in more detail now , it is hard to understand the intended behavior .
is "navigation in the main chat view/window" , the RHN ?  Do we need the navigation to be in sync with the navigation history and would  the back and forward keys  reflect the same list of navigated items , or would navigating change the order of the items on RHN, or should it have no affect when we navigate through the LHN on the RHN ? Do you thing we need a search history as mentioned at https://github.com/Expensify/App/issues/4612#issuecomment-1025119790 ?

**Comment by parasharrajat** [comment]
Agree, I will work on getting more details on the issue.

**Comment by mallenexpensify** [comment]
> Agree, I will work on getting more details on the issue.

Did you have a chance to dig in @parasharrajat ?

**Comment by parasharrajat** [comment]
I am trying to https://github.com/Expensify/App/issues/8657#issuecomment-1111015235.

**Comment by mallenexpensify** [comment]
@parasharrajat I put https://github.com/Expensify/App/issues/8657 on hold, with the plan being to fix this one first. Do you think that's the best approach?

**Comment by parasharrajat** [comment]
Technically both can be fixed together if the same solution goes to both and both issues can be fixed together if their solution does not overlap. The only advantage of fixing one by one is that we can focus on one thing first.

**Comment by mallenexpensify** [comment]
Let's see if we can get any takers for this specific issue/problem now then go from there.  Any reason I shouldn't double this to $16,000 @NikkiWines, @parasharrajat ?

**Comment by NikkiWines** [comment]
Sure, sounds good @mallenexpensify 👍

**Comment by mallenexpensify** [comment]
Doubled price to $16000

**Comment by ahmdshrif** [comment]
# Proposal 

the issue for LHN is different from RHN (search Screen) so this proposal is only for LHN 

## Problem :


https://user-images.githubusercontent.com/21364901/166839179-0c24bf24-3c57-4337-8647-40ddde7d95f5.mov

when we tap on LHN on user1 -> user2 -> user3 -> user4 -> user5 
then click back or use cmd + [  to go back 
the app will navigate to user3 -> user1

so the history is skipping one user from history. 
we can see this on video. 

##### Note here i add console.log(history) on this line and that what log on video https://github.com/Expensify/App/blob/cc38098efa7c35daef2a7e7cb88872d16c952188/src/libs/Navigation/CustomActions.js#L114
## cause  : 

when we calculate the history to push it to reset state as shown here :
https://github.com/Expensify/App/blob/cc38098efa7c35daef2a7e7cb88872d16c952188/src/libs/Navigation/CustomActions.js#L104-L113

we only push the drawer to history if we not have   drawer history  or we are not on drawer 
https://github.com/Expensify/App/blob/cc38098efa7c35daef2a7e7cb88872d16c952188/src/libs/Navigation/CustomActions.js#L106

### when we navigate from user1 to user2 
we are on drawer but we have drawer history  so we will push drawer open to history . 

### but when we navigate form user2 to user3  

we already have drawer history from last step  so we will not push any thing to history . and will keep the history length equal to the previous history length !!  but will replace all history with type: 'route' not drawer 

https://github.com/Expensify/App/blob/cc38098efa7c35daef2a7e7cb88872d16c952188/src/libs/Navigation/CustomActions.js#L100-L101


again form user3 -> user4 push drawer  
from user4 -> user5 not push any thing because we have drawer history and just replace old history with type route . 

and loop will keep going as shown on video  



## Sloution : 

we  can push drawer t close as we do before when we were on the drawer.  and we has drawer history with state open . 

 that is can be  like the 

```diff 
-        const hasDrawerhistory = _.find(state.history || [], h => h.type === 'drawer');
+        const hasCloseDrawerhistory = _.find(state.history || [], h => h.type === 'drawer' && h.status === 'closed');
-        if (!hasDrawerhistory || currentState.type !== 'drawer') {
+       if (!hasCloseDrawerhistory || currentState.type !== 'drawer') {
```

or if you prefer not to push the drawer as open since it's our default state we can push route. 
like this on line 113
```ts
        } else {
            history.push(screenRoute);
        }
```

## Result
 
as shown in the video it's now working fine.

https://user-images.githubusercontent.com/21364901/166842980-3a71eeee-72f9-45cf-a57e-81850dba66b3.mov


not I also tested the back button issue which was fixed by the code I edit on my proposal  and it work fine.

**Comment by parasharrajat** [comment]
Thanks @ahmdshrif I will review this as soon as possible.

**Comment by mallenexpensify** [comment]
@parasharrajat can you prioritize this since it's a $16k issue? I'd prefer to not have another proposal submitted before your review.

**Comment by parasharrajat** [comment]
ETA for review is 18 hours.

**Comment by parasharrajat** [comment]
Take a look at all the proposals now. Also, cross-referencing other issues due to overlapping proposals. For clarity, I am looking at [this](https://github.com/Expensify/App/issues/4612#issuecomment-1070032811) information in a proposal.

**Comment by ahmdshrif** [comment]
@parasharrajat I think my last  proposal cover this points . If you thing some point is not clear please mention it .

**Comment by parasharrajat** [comment]
- https://github.com/Expensify/App/issues/4612#issuecomment-1121164026 
- https://github.com/Expensify/App/issues/4612#issuecomment-1125057157
- https://github.com/Expensify/App/issues/8657#issuecomment-1107802797

All of the above proposals closely resemble. They work around the same root cause. I am testing them one by one now. They claim to fix both https://github.com/Expensify/App/issues/4612 and https://github.com/Expensify/App/issues/8657

**Comment by parasharrajat** [comment]
Both of the above proposals on this issue have the same problem. It takes two clicks to open the Chat when we resize the desktop to mobile view.

**Comment by ahmdshrif** [comment]
@parasharrajat  if we use the second option I mention it will work without this issue . we should push the route not the drawer in this case. 


> or if you prefer not to push the drawer as open since it's our default state we can push route.
like this on line 113 
```
        } else {
            history.push(screenRoute);
        }
```

## Result :

https://user-images.githubusercontent.com/21364901/168952792-f39a7835-cc62-4739-8c85-671e334632fd.mp4

**Comment by parasharrajat** [comment]
Why are we pushing this history entry? It is like mocking history, I do not agree with this change.

**Comment by ahmdshrif** [comment]
because we have a new route pushed to the drawer. 
if you use `navigate()` that will push automatically . since we user reset and add a new route we should add it to history also .

why you do not agree!!

**Comment by parasharrajat** [comment]
> because we have a new route pushed to the drawer.
if you use navigate() that will push automatically . since we use reset and add a new route we should add it to history also.

I will check it.

**Comment by parasharrajat** [comment]
Currently, we are evaluating the status of navigation issues in our app. It is decided that fixing issues upstream is the best path forward. So if an issue can be fixed upstream then it would be preferred over others. 

If you think that this is an issue with our repo, then please share a thorough explanation of it with your proposals. 

--- 

Although there are two different issues for LHN and RHN, I consider RHN and LHN behavior to be tightly dependent. It might be possible that fixing the RHN issue requires changing the code in our custom logic (if this is an issue with our logic).  Which could lead to a waste of resources on this issue. It is found over time that changing the custom Logic creates more problems. We would like to move away from that if possible.


For now, let's focus on the issues first to understand them better before finding a solution. I would say it would be better to get the root cause of the RHN issue as well. 

@ahmdshrif It seems that your solution is working but we are still not convinced if this is an issue with our logic.

**Comment by ahmdshrif** [comment]
Yes, we totally agree we should fix any issue upstream not workaround untill there is a limitation. 

but I see the LHN is happening because of the issue with our actions. it's different from other issues which may need fixing on react-navigation itself. 

why it's an issue from our side, not react-navigation?

because it works fine until we add this condition  
https://github.com/Expensify/App/blob/cc38098efa7c35daef2a7e7cb88872d16c952188/src/libs/Navigation/CustomActions.js#L106

this fixes another issue but it has broken the length of the history. 
If we go inside this condition we will increase the history by one. otherwise, we will keep the length as it's. 

react-navigation here says if we keep the history with the same length it will replace the previous route with the current route.
and that's what happened.  (check the log in the video on my main proposal ) 

![Screen Shot 2022-05-24 at 7 47 31 AM](https://user-images.githubusercontent.com/21364901/169958028-572b671d-895f-41f8-bdf3-bedf49ca2050.png)

https://github.com/react-navigation/react-navigation/blob/5b13f818f3c701ce33c2a30f5dd96763069556e4/packages/native/src/useLinking.tsx#L604-L605

you may say we push the drawer not route in the previous code. 
in both cases, react-navigation will push all-state (including the parameter ) with the state.
so we also make some mistakes when we push the state of a drawer in case we need to push the route, it was working because all state is pushed.

![Screen Shot 2022-05-24 at 7 58 41 AM](https://user-images.githubusercontent.com/21364901/169959518-16f7d5c5-1b19-45e8-a402-a754b40eeadb.png)

https://github.com/react-navigation/react-navigation/blob/5b13f818f3c701ce33c2a30f5dd96763069556e4/packages/native/src/useLinking.tsx#L579


In the end, react-navigation defines the history as the list of previously visited route keys and drawer open status
when we do push 3 routes to the history how we will expect 4 histories??
![Screen Shot 2022-05-24 at 8 25 12 AM](https://user-images.githubusercontent.com/21364901/169963452-93b46966-0f6c-46e8-bbd2-4f8b0162ada4.png)
https://github.com/ahmdshrif/react-navigation/blob/main/packages/routers/src/DrawerRouter.tsx#L43


cc: @parasharrajat

**Comment by marcaaron** [comment]
@mallenexpensify putting this issue on hold as well.

**Comment by ahmdshrif** [comment]
@parasharrajat   can you check this one https://github.com/Expensify/App/issues/8841#issuecomment-1159520598 
someone use the same [proposal](https://github.com/Expensify/App/issues/4612#issuecomment-1129534740) that I mention here to fix another issue and that gets merged and fixes this issue !!!

I mention this root cause and this fix before it. but this one gets hold.

cc: @mallenexpensify

**Comment by marcaaron** [comment]
That PR looks like it should maybe not have been sent 😬 and is a bit of a hack. At least I would have stopped it if I had seen it.

**Comment by ahmdshrif** [comment]
I think using dimension can be a workaround.
but push the history (which fixes this one)  I don't think . can you please check my explanation [here](https://github.com/Expensify/App/issues/4612#issuecomment-1135465924) why it's not a workaround in the LHN case.

**Comment by parasharrajat** [comment]
Thanks, @ahmdshrif for bringing this up. As mentioned by Marc https://github.com/Expensify/App/issues/8841#issuecomment-1194855072, we decided to deprecate the current implementation of CustomActions that's why your proposal was not accepted at that time and this issue was put on hold. 

we are yet to define a clear plan to solve navigation issues. Other PR will be reverted as well.

**Comment by NikkiWines** [comment]
Not overdue, issue is on hold

**Comment by NikkiWines** [comment]
Not tied to a regression, https://github.com/Expensify/App/issues/10770 is a similar issue and they're discussing closing it in favor of this issue.

**Comment by NikkiWines** [comment]
Unassigning myself while these are on hold since there's nothing actionable at this time

**Comment by JmillsExpensify** [comment]
Snagging myself now that I'm on the tracking issue.

**Comment by JmillsExpensify** [comment]
See the tracking issue for the latest updates.

**Comment by JmillsExpensify** [comment]
@marcaaron @puneetlath I wanted to quickly confirm that we still we do commit to doing this in the react navigation project?  I think largely the answer should be yes given how standard of a shortcut this is. However, I'm wavering a bit overall on the priority of keyboard actions overall. They all feel best done with a holistic, concerted push rather than piecemeal. Curious for your thoughts!

**Comment by JmillsExpensify** [comment]
Actually....wait a minute. How is this issue not an exact of the older, and more through https://github.com/Expensify/App/issues/8657. I think we should close out this issue as a dupe of that one.

**Comment by marcaaron** [comment]
I think we can incorporate it into the detailed design doc. My feeling about this issue is that the shortcuts never really worked as "expected" because, well, nobody ever promised that it would work a certain way. There are a lot of misunderstandings about how the memory history implemented in `react-navigation` works (especially with the drawer navigator in the mix).

There's like 250+ comments I am not going to revisit here, but `command + [` is the same thing as a browser back button. So if the browser back button works as expected on web and we switch to a simple stack navigator then I am pretty confident these shortcuts will "just work".

**Comment by marcaaron** [comment]
Also I think that, yes, the issues are very similar. If we switch to a stack navigator and include chats in the stack then it will be easier to reason about the history and expectations WRT shortcuts like `command + [` or `command + left arrow`.

**Comment by ahmdshrif** [comment]
>  the same thing as a browser back button. So if the browser back button works as expected on web and we switch to a simple stack navigator then I am pretty confident these shortcuts will "just work".

@marcaaron  same issue on the web. with the back and forward button 
when I work on proposals here I debug on the web most of the time. (IDK if you fix something now or not ) 
also, I should mention that we overwrite the history on navigation custome actions so this is the important part.

**Comment by marcaaron** [comment]
>I should mention that we overwrite the history on navigation custome actions so this is the important part.

Yes I added the code for it so I'm aware how it works 😄

**Comment by JmillsExpensify** [comment]
Ok cool, appreciate the additional context @marcaaron. So where do we do from here? Give that this is a duplicate of an older issue, I'm still inclined to close this one. The older issue also has better reproduction steps to boot. In fact, I'll go ahead and do so now, though any discussion can keep happening if required.

**Comment by michaelhaxhiu** [comment]
https://expensify.slack.com/archives/C02NK2DQWUX/p1670871808439109?thread_ts=1668807800.467689&cid=C02NK2DQWUX
